test/performance/odp_timer_perf.c

```diff
@@ -805,6 +824,130 @@ static int set_cancel_mode_worker(void *arg)
 	return ret;
 }
 
+static int set_expire_mode_worker(void *arg)
+{
+	int status, thr;
+	uint32_t i, j, exit_test;
+	odp_event_t ev;
+	odp_timeout_t tmo;
+	uint64_t c2, diff, nsec, time_ns, target_ns, period_tick, current_tick;
+	odp_timer_t timer;
+	odp_timer_pool_t tp;
+	odp_timer_start_t start_param;
+	odp_time_t t1, t2;
+	time_stat_t before, after;
+	timer_ctx_t *ctx;
+	thread_arg_t *thread_arg = arg;
+	test_global_t *global = thread_arg->global;
+	test_options_t *opt = &global->test_options;
+	uint32_t num_tp = opt->num_tp;
+	uint64_t cycles = 0;
+	uint64_t events = 0;
+	uint64_t rounds = 0;
+	uint64_t c1 = 0;
+	int meas = 1;
+	int ret = 0;
+
+	memset(&before, 0, sizeof(time_stat_t));
+	memset(&after, 0, sizeof(time_stat_t));
+
+	thr = odp_thread_id();
+
+	/* Start all workers at the same time */
+	odp_barrier_wait(&global->barrier);
+
+	t1 = odp_time_local();
+
+	while (events < opt->test_rounds * opt->num_timer / opt->num_cpu) {
+		if (meas) {
+			c1   = odp_cpu_cycles();
+			meas = 0;
+		}
+
+		ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+		rounds++;
+
+		exit_test = odp_atomic_load_u32(&global->exit_test);
+		if (odp_likely(ev == ODP_EVENT_INVALID && exit_test < num_tp))
+			continue;
+
+		c2      = odp_cpu_cycles();
+		diff    = odp_cpu_cycles_diff(c2, c1);
+		cycles += diff;
+
+		if (ev == ODP_EVENT_INVALID && exit_test >= num_tp)
+			break;
+
+		events++;
+		meas = 1;
+		tmo = odp_timeout_from_event(ev);
+		ctx = odp_timeout_user_ptr(tmo);
+		i = ctx->tp_idx;
+		j = ctx->timer_idx;
+		timer = global->timer[i][j];
+		tp = global->timer_pool[i].tp;
+		period_tick = global->timer_pool[i].period_tick;
+		current_tick = odp_timer_current_tick(tp);
+		time_ns = odp_timer_tick_to_ns(tp, current_tick);
```

Comment:
I think time API should be used for accuracy measurement (as the other test case does). Time API is likely more accurate than timer ticks (and has strict version).

Reply:
v3.